### ISSUE TITLE
fix(model/v41): fail closed at V4.1 forward entrypoints, not only loader/selection

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12967]
﻿

# fix(model/v41): fail closed at V4.1 forward entrypoints, not only loader/selection

```routing
lane: model
paths: ["internal/model/forward.go", "internal/model/kv.go", "internal/model/v41_failclosed_test.go"]
expected_steps: 3
repo: fak
public_issue: anthony-chaudhary/fak#12640
```

## Parent context
Parent: #12640 (native DeepSeek V4.1 Flash execution and physical qualification).

## Current state
The V4.1 fail-closed invariant now covers every loader (AWQ/EXL2/GPTQ/safetensors/quant, #12891/#12933/#12902)
and forward SELECTION (`ClassifyForwardPath`, #12902). It does NOT cover the direct in-memory forward
entrypoints: on `origin/main` `bc9778ea8`, `git grep -n 'IsDeepSeekV41\|refuseDeepSeekV41' -- internal/model/forward.go internal/model/kv.go`
returns nothing. `Model.Forward` (`forward.go:461`), `Session.Prefill` (`kv.go:1016`) and `Session.Step`
(`kv.go:1265`) would run the generic `layer()`/token loop over V4.1 geometry with no typed refusal if a
V4.1 `*Model` were ever constructed in memory (e.g. by a future native leaf, a test helper, or a
deserialized state that bypasses the loaders).

## Problem frame
- Centrality: Core - the entrypoint is the last unguarded seam before V4.1 geometry reaches a generic forward.
- P1: advanced - closes the fail-closed invariant at the forward boundary, not only load and selection.
- P2: advanced - prevents silent partial logits from partial V4.1 work landing out of order.
- P3: preserved - non-V4.1 families are unchanged by an IsDeepSeekV41-gated refusal.
- P4: advanced - negative admission evidence at the forward entrypoints.
## Why now
#12902's acceptance criterion "a partially implemented forward stage returns a typed error, not partial
logits" is satisfied at selection and load, but not at the entrypoints themselves. As the forward leaves
(#12901, #12895-#12899) land out of order, an unguarded entrypoint is exactly the path by which partial
V4.1 work silently produces logits. Recorded as a discovered edge case from #12902.

## Core through-line
Add a shared `Session`/`Model` V4.1 refusal (mirroring `requirePreNorm`/`requireGLMDsaSession`) that
fails closed with `ErrV41NativeUnsupported` at `Model.Forward`, `Session.Prefill` and `Session.Step`,
plus a negative test proving a hand-built V4.1 config cannot reach the generic loop. Do not weaken
existing rejections and do not change any public loader signature.

## Working spine
Intake -> hand-built V4.1 config reaches an entrypoint -> add typed refusal at Forward/Prefill/Step ->
negative witness proves no partial logits -> landed receipt.

## Gold-plating boundary
No new forward capability. Guard-and-test only. No loader, admission, or ABI change. Do not remove the
guard once a native V4.1 forward lands; replace it with the real forward in the owning leaf.

## Witness
```bash
go -C ..\fak test ./internal/model -run 'TestV41FailClosedForwardEntrypoints' -count=1
```

### File:Line Seam
`internal/model/forward.go:461` (`Model.Forward`), `internal/model/kv.go:1016` (`Session.Prefill`),
`internal/model/kv.go:1265` (`Session.Step`).

### Blast Radius & Affected Lanes
Primary lane `model`; non-V4.1 families unchanged (the guard is `IsDeepSeekV41()`-gated).

### Quarantined Fallback Mechanism
The typed error is the quarantine; trunk stays green with V4.1 execution held. If a guard cannot be
placed at an entrypoint without disturbing a hot path, fail closed at the earliest reachable common
seam instead and record the deviation in the closing comment.

## Scoped Acceptance Criteria
- [ ] [SW-VERIFIED] A V4.1 config reaching `Model.Forward` returns/panics with the typed `ErrV41NativeUnsupported`, not partial logits.
- [ ] [SW-VERIFIED] A V4.1 `Session.Prefill` and `Session.Step` fail closed identically.
- [ ] [SW-VERIFIED] Non-V4.1 families are byte-for-byte unchanged.
- [ ] [SW-VERIFIED] No test is skipped or weakened.

## Done condition
The scoped acceptance criteria pass with the cited witness; `go test ./internal/model -count=1`, vet and
the 5-gate boundary check are clean.

## Acceptance gate
Focused witness passes, package tests are green, vet and 5-gate boundary checks are clean.

## Work estimate
Estimate: 2 points (bounded leaf).

## Overall completion contribution
Contribution: 2/40 points (provisional baseline).

## Completion standard
development

## Closure binding
The resolving commit cites this issue with Fixes/Ref and carries the repository-qualified lane trailer.

**Comment by anthony-chaudhary** [comment]
﻿Landed on origin/main at **e1071e739c7f84c556cbb61c1af6f786e1b3c613** ("fix(model/v41): fail closed at V4.1 forward entrypoints, not only loader/selection (fak model)"), Fixes #12967.

## Witness (on the landed tree)
```text
go test ./internal/model -run 'TestV41FailClosedForwardEntrypoints' -count=1 -v  => PASS (4 subtests)
go test ./internal/model -run 'TestV41|TestDeepSeekV41|TestV4' -count=1          => ok 2.486s
go test ./internal/model -count=1                                                => ok 157.072s
go vet ./internal/model                                                          => exit 0
go run ./cmd/fak-boundary check -fak-dir <public worktree> -all                  => PASS (0 violations, 5 gates)
```

## Change
Added 
equireV41NativeUnsupported on Model (forward.go) and Session (kv.go), called at Model.Forward, Session.Prefill and Session.Step. A V4.1 *Model constructed directly in memory (bypassing all loaders/selection) now fails closed with the typed ErrV41NativeUnsupported instead of reaching the generic layer()/token loop and emitting partial logits. Non-V4.1 families are IsDeepSeekV41-gated and unchanged. Test-only coverage in 41_failclosed_test.go.

[SW-VERIFIED] component/entrypoint refusal only; no native V4.1 execution claim. Native V4.1 remains open in #12640.
